cs/src/Connections/TunnelConnection.cs

```diff
@@ -249,18 +283,22 @@ protected async Task<bool> RefreshTunnelAccessTokenAsync(CancellationToken cance
             TunnelAccessTokenProperties.GetTokenTrace(this.accessToken));
         try
         {
-            var newTunnelAccessToken = await GetFreshTunnelAccessTokenAsync(cancellation);
-            if (!string.IsNullOrEmpty(newTunnelAccessToken))
+            if (!await OnRefreshingTunnelAccessTokenAsync(cancellation))
             {
-                TunnelAccessTokenProperties.ValidateTokenExpiration(newTunnelAccessToken);
-                Trace.TraceInformation(
-                    "Refreshed tunnel access token. New token: {0}",
-                    TunnelAccessTokenProperties.GetTokenTrace(newTunnelAccessToken));
-                this.accessToken = newTunnelAccessToken;
-                return true;
+                return false;
             }
 
-            return false;
+            // Access token may be null if tunnel allows anonymous access.
+            if (this.accessToken != null)
+            {
+                TunnelAccessTokenProperties.ValidateTokenExpiration(this.accessToken);
+            }
+
+            Trace.TraceInformation(
+                "Refreshed tunnel access token. New token: {0}",
```

Comment:
Avoid logging the full token, especially in non-verbose logs. I see this line was moved from above, but still...

Reply:
Logging verbose now.